doc/program_services.md

```diff
@@ -0,0 +1,1175 @@
+7. The called program restores GR14 to GR12 from that save area (restores registers at entry); there are cases when GR15, GR0, and/or GR1 are not restored -- depends on the program. 
+8. The called program returns to the caller by doing an unconditional branch to the address in GR14.
+
+This calling convention is explained in detail in the [IBM z/OS Version 2 Release 4 MVS Programming: Assembler Services
+Guide (SA23-1368-40)](https://www-01.ibm.com/servers/resourcelink/svc00100.nsf/pages/zOSV2R4sa231368/$file/ieaa600_v2r4.pdf). Refer to Chapter 2 - Linkage conventions.
+
+GR15 is expected to contain a return code upon exit by convention.
+
+## Program parameters
+
+There are two methods of passing parameters to a program.
+
+* At compile time using SYSPARM
+* At run time using PARM
+
+In either case enclose the entire parm to be passed in double quotes.
+
+The double quotes are required to handle commas and spaces that otherwise cause command processors to split the parm.
+
+The double quotes are not required if there are no commas or spaces in the text. 
+
+### Compile time - SYSPARM
+
+If single quotes are included in text, they are passed on to &SYSPARM.
+
+There is a limit of 32767 bytes for the text. 
+
+SYSPARM will be transferred to the macro variable &SYSPARM.
+
+``` dos
+mz390 ... "SYSPARM(HELLO WORLD)"
+```
+Access the text by coding:
+
+``` hlasm
+label    DC    C'&SYSPARM'
+...... will translate to 
+label    DC    C'HELLO WORLD'
+```
+
+### Run time - PARM
+
+PARM can be accessed via GR1 at program entry.
+
+GR1 will contain an address that points to a fullword in storage which points to a storage location with the parm length in a halfword followed by the parm value.
+
+``` dos
+ez390 ... "PARM(HELLO WORLD)"
+```
+
+The fullword pointed to by GR1 points to:
+
+``` hlasm
+         DC   H'11',C'HELLO WORLD'`
+```
+
+If single quotes are included around text in PARM they are removed.
+
+``` dos
+ez390 .... "PARM('HELLO WORLD')" 
+```
+
+Will also result in the fullword pointed to by GR1 pointing to:
+
+``` hlasm
+         DC   H'11',C'HELLO WORLD'`
+```
+
+#### Usage
+
+The program is run with the parameter passed
+
+``` dos
+ez390 myprog "PARM('HELLO THERE')"
+```
+
+The program code:
+
+``` hlasm
+         L     R2,0(,R1)       R2=address of parm length(HW)+value
+         LH    R3,0(,R2)       R3=length of parm value
+         XGR   R0,R0           Set R0 to zero (for MVCOS instruction)
+         MVCOS PARM,2(R2),R3   Move R3 bytes at 2(R2) to PARM
+```
+* GR1 is populated by ez390 with the parm address pointer.
+* GR2 is loaded with the address of the parm length and address - fullword pointed to by GR1.
+* GR3 is loaded with the parm length - halfword value pointed to by GR2.
+* GR3 would contain decimal 11 which is the length of the parm "HELLO THERE".
+* The actual parm value of "HELLO THERE" begins 2 bytes past the address in GR2.
+* The `MVCOS` instruction will move the parm to the storage at label PARM.
```

Comment:
While this is technically correct, the MVCOS instruction is probably unknown to most users (I had to look it up). A more classical example would be something like

```
         L     R2,0(,R1)          R2 = address of the "PARM=" value  area
         LH  R3,0(,R2)          R3 = length of the parm value
         BCTR  R3,0             Convert length to length code
         EX    R3,MOVE       Copy value to work area
         ...
MOVE     MVC   PARM(0),2(R2)   Executed instruction
PARM     DC    CL80' '        Work area for parm value
```
I'm sure you can clean up what I've tried to enter as code here and  update the six bullets. The first four are okay. 

Insert a new fifth bullet
"- GR3 is decremented by 1, making it a length code."

Replace the last bullet with something like

"- The "EX" instruction executes the instruction at label MOVE which will move the parm to the storage at label PARM."

Would be nice to shorten this bullet so that it fits on one line.

Reply:
OK.. so my assembler skill level is pretty low so just go with me here... as a learner..

1. Why would you use a branch on count to do a subtract when you could do a subtract (Am I missing something??).

2. The EX instruction is not that clear to me, but possibly obvious to a seasoned assembler programmer. I had to read the POP for at least 10 minutes trying to work out how that does a move.
```
EX       R3,MOVE  
......
MOVE     MVC   PARM(R0),2(R2)
```
So EX will execute the instruction located at MOVE address, but change the first parm after the opcode with the value in R3.... 

In this case, the MVC instruction as hex is `D2*00*F02C2002` and the part between the * is updated with the value in R3.

You need to know the fact that the MVC L1 parameter will be one updated by the EX instruction. 
You also need to know that in the machine code length (length code) is length-1.
This may not be obvious to a beginner coder, especially when they are reaching the manual to learn.

My example seemed obvious.. it's just not a well used instruction.

If you want to use that example, I think the comments need to be updated..

Insert a new fifth bullet
* GR3 is decremented by 1, making it a length code **(length code = length-1 as used by machine code)**.

Replace the last bullet with something like

* The "EX" instruction executes the instruction at label MOVE which will move the parm to the storage at label PARM **using length from R3** 

Let me know your thoughts based on my feedback and I will update accordingly..

Reply:
As I saw in the email thread, most of this has been talked about. Here's my 2 cents worth.

The instruction at label MOVE is "MVC   PARM(0),2(R2)", not "MVC   PARM(R0),2(R2)". The zero in the instruction is the length being moved, so will occupy the "length" position in the generated instruction. A more explicit example better explains this. Consider "MVC   4(8,R1),5(R2)". This moves the 8 bytes that start at 5(R2) to the 8 bytes starting at 4(R1). The encoded instruction is X'D20710042005'. Note that the length, 8, has been converted to a "length code" -- one less than the length. IBM did this so that MVC could be used to move up to 256 bytes (256 = X'100'; 256-1=255=X'FF', which will fit in the length position of the encoded instruction). One little quirk:
 "MVC   4(0,R1),5(R2)" is encoded as X'D20010042005' -- the 0 length is not decremented.

Suppose that we want to move an unknown number of bytes (but no more than 256) from 5(R2) to 4(R1) using MVC. The number of bytes to move is in R3. The classical technique from System 360 days is to convert the length to a length code, then use an execute instruction:
```
         BCTR  R3,0
         EX    R3,MOVE
         ...
MOVE     MVC   4(0,R1),5(R2)
```

The EX instruction does a logical OR of the low byte of R3 with bits 8-15 of the target instruction, which iswhere the length code is stored. Since the value there is 0, this effectively stores the low byte of R3 in those 8 bits, resulting in the length code becoming the desired length code, so the executed MVC moves the correct number of bytes.

Since R3 originally contained the length, it had to be decremented before the EX instruction was executed. Yes, one could have used something like "S     R3,=F'1'" to decrement R3, but that is a 4-byte instruction and a 4-byte literal, using up 8 bytes of memory, whereas "BCTR  R3,0" only uses 2 bytes and does the same thing. Remember, this example is from the 1960s when memory was a scarce resource.

I used this technique up until I retired in 2014. The only "improvement" I did was replace the executed instruction with something like

MOVE     MVC   4(*-*,R1),5(R2)
or
MOVE     MVC   PARM(*-*),2(R2)

"*-*" is 0, but it makes it clear that the length value will be supplied elsewhere, most likely via an execute instruction.

As was pointed out in the email thread, there are many ways to copy the parm value. If you feel strongly about leaving it with the MVCOS, then do so.

Reply:
Have changed to use BCTR/EX in ecb29a8.